proxy/pop3_test.go

```diff
@@ -0,0 +1,302 @@
+// pop3_proxy_test.go - pop3 proxy tests
+// Copyright (C) 2017  David Anthony Stainton
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package proxy
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/textproto"
+	"os"
+	"strconv"
+	"sync"
+	"testing"
+
+	"github.com/boltdb/bolt"
+	"github.com/katzenpost/client/constants"
+	"github.com/katzenpost/client/storage"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	testUser = "alice"
+	testPass = "teatime475"
+)
+
+func setupPop3Db(dbFile, bucketName string) error {
```

Comment:
Also worth setting up a utility client, if there isn't one, to attempt connecting this pop3 service against a remote server, so that a user can provide diagnostics on why it fails against their provider, which may handle it's pop3 server implementation differently from the very scripted one used in this file.

Reply:
What? I think you might be mistaken about our design here. The SMTP and POP3 services are only run locally on the client endpoint device. The "Providers" will not be running any SMTP or POP3 services at all. Instead we've implemented a custome wire protocol that can handle all of that. See https://github.com/Katzenpost/docs/blob/master/specs/end_to_end.txt and https://github.com/Katzenpost/docs/blob/master/specs/wire-protocol.txt

Reply:
I guess: what's under test here - it looks like you're giving a very scripted conversation here that won't always handle real-world diversity.  If it's the service that's being emulated, then you should have a test-harness able to make sure it's good with a variety of real clients.

Reply:
for this grant project it only has to work with k9mail... but yeah you are right that we should test with all the e-mail clients that we want it to work with. yawning wrote this pop3 library... and followed the rfc very carefully so i think this kind of testing is good enough for now.